content/blog/gpu-hackathon.md

```diff
@@ -0,0 +1,55 @@
+
+With our mentors we started profiling, only to find that we had a major problem with lots of small `cudaMalloc` and `cudaMemcpy` calls rather than doing one big one and dividing that up among the threads. This lead to performance that was considerably slower on the GPU than the CPU and took a day's refactoring to fix.  It also turns out that having lots of small allocations and copies renders the profiler almost unusable so we could not really begin to optimise our "kernel" (the bit of code we were trying to run on the GPU).
+
+With that sorted out, we explored a variety of optimisations for improving our volatility model, focussing on `double` vs `float` maths (with this making a significant difference on the GPU, wheras I'd always thought of `float` as essentially an anachronism on the CPU). We found that as we increased the number of particles (independent simulations) and the number of steps that we ran for, the GPU performance increased from being 5x slower than the speed of the GPU to about 150x faster. It's likely that most of the metrics here are measuring the speed at which we can copy data from R, through a C++ later onto the device, and then reversing this to retrieve the results.
+
+Our SIRS model, which relies on samples from a binomial distribution, was more challenging to optimise, though closer to our real model and with more interesting calculations in the kernel. Our initial version was able to run at best 10x faster than the CPU (even after fixing our memory allocations) which we were a bit disappointed in because running the model on a 10-core system is fairly straightforward.  For the parameters we were using to start optimising, our kernel was taking ~2s to run, vs 1ms for our optimised version of the volatility model!
+
+With a lot of help from our mentors and one of the people who has worked on the NVIDIA profiler, we managed to track down the problem to the way we were dealing with divergence - adding `__syncwarp()` to bring threads that have come out of sync due to different numbers of iterations to successfully sample from the binomial distribution. Once we fixed that, our kernel was running in ~40ms and we were much happier. After a few of optimisation (moving to `float` rather than `double` and removing branches that were optimisations in CPU code but caused divergence on the GPU) we ended up with our kernel running in 15ms, a speedup of 130x.
```

Comment:
```suggestion
With a lot of help from our mentors, and one of the people who has worked on the NVIDIA profiler, we managed to track down the problem. Due to the way we were dealing with divergence we needed to add `__syncwarp()` to bring threads that have come out of sync due to different numbers of iterations to successfully sample from the binomial distribution. Once we fixed that, our kernel was running in ~40ms and we were much happier. After a few more rounds of optimisation (moving to `float` rather than `double` and removing branches that were optimisations in CPU code but caused divergence on the GPU) we ended up with our kernel running in 15ms, a speedup of 130x.
```